File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.js
/**
 * @param {number[][]} heights
 * @return {number[][]}
 */
var pacificAtlantic = function(heights) {
    let rows = heights.length
    let cols = heights[0].length
    let res = []
    let pac = new Array(rows).fill().map(() => new Array(cols).fill(0))
    let atl = new Array(rows).fill().map(() => new Array(cols).fill(0))

    //left & right
    for(let row = 0; row<rows; row++){
        dfs(row, 0, pac, heights[row][0])
        dfs(row, cols-1, atl, heights[row][cols-1])
    }
    //top & bottom
    for(let col = 0; col<cols; col++){
        dfs(0, col, pac, heights[0][col])
        dfs(rows-1, col, atl, heights[rows-1][col])
    }

    function dfs(row, col, ocean, prevHeight){
        if(row < 0 || row >= rows || col < 0 || col >= cols) return;
        if(prevHeight > heights[row][col]) return;
        if(ocean[row][col] == 1) return;
        ocean[row][col] = 1
        dfs(row + 1, col, ocean, heights[row][col])
        dfs(row - 1, col, ocean, heights[row][col])
        dfs(row, col + 1, ocean, heights[row][col])
        dfs(row, col - 1, ocean, heights[row][col])
    }

    for(let row=0; row < rows; row++){
        for(let col=0; col<cols; col++){
            if(pac[row][col] == 1 && atl[row][col] == 1){
                res.push([row,col])
            }
        }
    }
    return res
};

File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        rows = len(heights)
        cols = len(heights[0])

        pacific = set()
        atlantic = set()

        def dfs(r, c, visited, prev_height):
            if (r,c) in visited or heights[r][c] < prev_height:
                return

            visited.add((r,c))

            for dr, dc in [(0,1),(0,-1),(1,0),(-1,0)]:
                nr = r + dr
                nc = c + dc
                if 0 <= nr < rows and 0 <= nc < cols:
                    dfs(nr, nc, visited, heights[r][c])
            
        
        for r in range(rows):
            dfs(r, 0, pacific, heights[r][0])
            dfs(r, cols-1, atlantic, heights[r][cols-1])
        
        for c in range(cols):
            dfs(0, c, pacific, heights[0][c])
            dfs(rows-1, c, atlantic, heights[rows-1][c])
        
        return list(map(list, pacific & atlantic))